File: Chapter 12 - Mediation and Instrumental Variables/chap 12 - analyzing data.py
# -*- coding: utf-8 -*-
"""
Created on Sun May 10 11:44:30 2020

@author: Florent
"""

#################################
##### This script analyzes the data used in chapter 11, 
##### Mediation and Instrumental Variables
#################################

##### Setup #####

#Common libraries
import pandas as pd
from statsmodels.formula.api import ols
import numpy as np


#Loading the data from the chapter on moderation
hist_data_df = pd.read_csv('chap10-historical_data.csv')

##### 1. Mediation #####

#Regressions used in the text
ols("duration~play_area", data=hist_data_df).fit().summary()
ols("groceries_purchases~play_area", data=hist_data_df).fit().summary()
ols("groceries_purchases~duration", data=hist_data_df).fit().summary()
ols("groceries_purchases~duration+play_area", data=hist_data_df).fit().summary()

### Bootstrap CI for percentage mediated 

def percentage_mediated_fun(dat_df):
    
    total_effect = ols("groceries_purchases~play_area", data=dat_df).fit(disp=0).params['play_area']
    coeff_med1 = ols("duration~play_area", data=dat_df).fit(disp=0).params['play_area']
    coeff_med2 = ols("groceries_purchases~duration", data=dat_df).fit(disp=0).params['duration']
    mediated_effect = coeff_med1 * coeff_med2
    percentage_mediated = mediated_effect / total_effect
    return percentage_mediated 

percentage_mediated_fun(hist_data_df)

def boot_CI_fun(dat_df, metric_fun, B = 100):
  #Setting sample size
  N = len(dat_df)
  conf_level = 0.9
  coeffs = []
  
  
  for i in range(B):
      sim_data_df = dat_df.sample(n=N, replace = True)
      coeff = metric_fun(sim_data_df)
      coeffs.append(coeff)
  
  coeffs.sort()
  start_idx = round(B * (1 - conf_level) / 2)
  end_idx = - round(B * (1 - conf_level) / 2)
  
  confint = [coeffs[start_idx], coeffs[end_idx]]  
  
  return(confint)
#boot_CI_fun(hist_data_df, percentage_mediated_fun)

##### 2. Instrumental variables #####

#Common libraries
import pandas as pd
from statsmodels.formula.api import ols
import numpy as np

#Library for IV regressions
from linearmodels.iv import IV2SLS

#Loading the experimental data from the chapter 9
exp_data_df = pd.read_csv('chap10-experimental_data.csv')

#Reformat group variable to binary in experimental data
exp_data_df.group = np.where(exp_data_df.group == 'treat', 1, 0)


#Reduced regression, coeff = 1.6
red_mod = ols("M6Spend~group+age+reason", data=exp_data_df).fit(disp=0)
red_mod.summary()

#First stage regression, coeff = 0.5
S1_mod = ols("call_CSAT~group+age+reason", data=exp_data_df).fit(disp=0)
S1_mod.summary()

#Baseline (biased) regression, coeff = 4.00
lm_mod = ols("M6Spend~call_CSAT+age+reason", data=exp_data_df).fit(disp=0)
lm_mod.summary()

#IV regression, coeff = 2.99
iv_mod = IV2SLS.from_formula('M6Spend ~ 1 + age + reason + [call_CSAT ~ group]', 
                             exp_data_df).fit()
iv_mod.params

File: Chapter 12 - Mediation and Instrumental Variables/chap 12 - analyzing data.R
#################################
##### This script analyzes the data used in chapter 12, 
##### Mediation and Instrumental Variables
#################################

##### Setup #####

# Common libraries
suppressMessages(suppressWarnings(library(tidyverse)))
library(boot) #Required for Bootstrap simulations
library(rstudioapi) #To load data from local folder
library(ggpubr) #To generate multi-plots

# Chapter-specific libraries
library(ivreg) # for IV regressions

### Setting the working directory to the parent folder of this script (Rstudio only)
sourceDir <- rstudioapi::getActiveDocumentContext()$path %>% str_extract("^.+/")
setwd(sourceDir)

set.seed(1234)
options(scipen=10)

#Loading the data from the chapter on moderation
hist_data <- read_csv("chap11-historical_data.csv")

#Formatting the data
hist_data <- hist_data %>%
  mutate(store_id = factor(store_id)) %>%
  mutate(day = factor(day)) %>%
  mutate(children = factor(children)) %>%
  mutate(play_area = factor(play_area))

##### 1. Mediation #####

#### 1.1. Definition and Measurement ####

summary(lm(duration~play_area, data=hist_data))
summary(lm(groceries_purchases~play_area, data=hist_data))
summary(lm(groceries_purchases~duration, data=hist_data))
summary(lm(groceries_purchases~duration+play_area, data=hist_data))

### Bootstrap CI for percentage mediated 
percentage_mediated_fun <- function(dat){
  total_effect <- summary(lm(groceries_purchases~play_area, data=dat))$coefficients['play_area1', 'Estimate']
  coeff_med1 <- summary(lm(duration~play_area, data=dat))$coefficients['play_area1', 'Estimate']
  coeff_med2 <- summary(lm(groceries_purchases~duration, data=dat))$coefficients['duration', 'Estimate']
  mediated_effect <- coeff_med1 * coeff_med2
  percentage_mediated <- mediated_effect / total_effect
  return(percentage_mediated)
}
percentage_mediated_fun(hist_data)

boot_CI_fun <- function(dat, metric_fun, B = 20){

  boot_metric_fun <- function(dat, J){
    boot_dat <- dat[J,]
    return(metric_fun(boot_dat))
  }
  boot.out <- boot(data=dat, statistic=boot_metric_fun, R=B)
  confint <- boot.ci(boot.out, conf = 0.90, type = c('perc'))
  CI <- confint$percent[c(4,5)]
  
  return(CI)
}
boot_CI_fun(hist_data, percentage_mediated_fun)

##### 2. Instrumental variables #####

#Loading the experimental data from chapter 9
exp_data <- read_csv("chap9-experimental_data.csv")

#Reformating the data
exp_data <- exp_data %>%
  mutate(center_ID = factor(center_ID)) %>%
  mutate(rep_ID = factor(rep_ID)) %>%
  mutate(reason = factor(reason)) %>%
  mutate(group = factor(group)) 

#Reduced regression, coeff = 1.6
red_mod <- lm(M6Spend~group+age, data=exp_data)
summary(red_mod)

#First stage regression, coeff = 0.5
S1_mod <- lm(call_CSAT~group+age, data=exp_data)
summary(S1_mod)

#Baseline (biased) regression
lm_mod <- lm(M6Spend~call_CSAT+age, data=exp_data)
summary(lm_mod)

### IV regression
iv_mod <- ivreg::ivreg(M6Spend~call_CSAT + age | group + age, data=exp_data)
summary(iv_mod)

iv_metric_fun <- function(dat){
  iv_mod <- ivreg::ivreg(M6Spend~call_CSAT + age | group + age, data=dat)
  iv_summ <- summary(iv_mod)
  coeff <- iv_summ$coefficients['call_CSAT','Estimate']
  return(coeff)
}
boot_CI <- boot_CI_fun(exp_data, iv_metric_fun)